server/src/main/java/org/elasticsearch/transport/RemoteClusterConnection.java

```diff
@@ -255,40 +253,52 @@ public String executor() {
     }
 
     /**
-     * Returns a connection to the remote cluster. This connection might be a proxy connection that redirects internally to the
-     * given node.
+     * Returns a connection to the remote cluster, preferably a direct connection to the provided {@link DiscoveryNode}.
+     * If such node is not connected, the returned connection will be a proxy connection that redirects to it.
      */
     Transport.Connection getConnection(DiscoveryNode remoteClusterNode) {
-        DiscoveryNode discoveryNode = connectedNodes.get();
+        if (connectedNodes.contains(remoteClusterNode)) {
```

Comment:
should we use `transportService .nodeConnected` instead. I mean if we for instance have the local cluster configured as a remote cluster we get this optimization as well?

Reply:
makes sense, great catch!